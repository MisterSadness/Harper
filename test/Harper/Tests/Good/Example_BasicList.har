value List a = {
  // Types can be simple (single-variant) or union (variant) types.
  variant Empty = {
    isEmpty :: Bool;
    isEmpty = true;
  };
  
  variant NonEmpty = {
    // Data visible publicly for deconstruction, as in:
    // Nonempty { head: val x, tail: val xs }
    data = {
      head :: a;
      tail :: List a;
    }
    
    isEmpty :: Bool;
    isEmpty = false;
  };
};

cons :: a -> List a -> List a;
cons a l = val NonEmpty { head = a, tail = l };

printList :: List a -> sideeffect -> ();
printList xs = {
    if xs.isEmpty {
      return;
    }

    // Print is a built-in function of type a -> sideeffect
    eval print xs.head ();
    eval printList xs.tail ();
};

main :: sideeffect -> ();
main = {
  eval printList (cons 42 | cons 17 | cons 37 (val Empty { })) ();
};