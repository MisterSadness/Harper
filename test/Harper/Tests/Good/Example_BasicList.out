421737

Execution ended with value:
()